eliot/_message.py

```diff
@@ -106,16 +124,57 @@ def write(self, logger=None, action=None):
         """
         if logger is None:
             logger = _output._DEFAULT_LOGGER
-        if action is None:
-            action = currentAction()
-        if action is None:
-            action = _defaultAction
-        contents = self._contents.copy()
-        contents["timestamp"] = self._timestamp()
-        if action is not None:
-            contents["task_uuid"] = action._identification["task_uuid"]
-            contents["task_level"] = action._nextTaskLevel().level
-        logger.write(contents, self._serializer)
+        logged = self._freeze(action=action)
+        logger.write(logged.as_dict(), self._serializer)
```

Comment:
OK, so, what you're doing here is:

Taking dictionary, copying it, mutating copy, then recreating it, for every logged message. Under high logging loads this is a whole bunch of additional unnecessary work.

Instead I suggest:

1. `Message._freeze` returns an updated `pmap` with additional entries added.
2. `WrittenMessage` is just a thin wrapper around a mapping, so:
    * `WrittenMessage.contents` just contain the `pmap` constructed by `_freeze`.
    * `WrittenMessage.task_level` and friends are properties calculated on demand.

This means this refactor adds much less additional work to the critical code path.

The bigger picture point is that once something has been made immutable (in this case via `_freeze`) you can trust it to remain immutable, no need to do unnecessary copies or transforms, you can trust it will never change behind your back.

Why do you need to return anything from `Message.write()`, BTW? Just for testing, or do you expect it to be part of the future parsing API?